app/app.go

```diff
@@ -158,20 +169,28 @@ func (app *MerkleEyesApp) doTx(tree merkle.Tree, tx []byte) abci.Result {
 
 func (app *MerkleEyesApp) Commit() abci.Result {
 
-	hash := app.state.Commit()
-
+	app.hash = app.state.Hash()
 	app.height++
+	fmt.Printf("height=%d,hash=%x\n", app.height, app.hash)
+
 	if app.db != nil {
-		app.db.Set(eyesStateKey, wire.BinaryBytes(MerkleEyesState{
-			Hash:   hash,
+		// This is in the batch with the Save, but not in the tree
+		tree := app.state.Append().(*iavl.IAVLTree)
```

Comment:
If we save the tree first, then crash, since the tree currently has no height information we will load H-1 as the state, and the replay with tendermint will double up the Sets/Removes for the last height H. If we batch it all together, the height and the tree are in sync, either they get in LevelDB together or they are lost.